### DISCUSSION TITLE
Embeddable relations serialization

### DISCUSSION THREAD

**Discussion by leonardbinet** [mikro-orm/mikro-orm#6171] [Q&A]
### Describe the bug

When serializing an embeddable, I would have expected the relations to be serialized as well. In below example, group name property isn't serialized.

### Reproduction

https://github.com/mikro-orm/mikro-orm/pull/6169

### What driver are you using?

@mikro-orm/postgresql

### MikroORM version

6.3.13

### Node.js version

v20.11.1

### Operating system

MacOS Sonoma 14.7 (23H124)

### Validations

- [X] Read the [Contributing Guidelines](https://github.com/mikro-orm/mikro-orm/blob/master/CONTRIBUTING.md).
- [X] Read the [docs](https://mikro-orm.io/docs).
- [X] Check that there isn't [already an issue](https://github.com/mikro-orm/mikro-orm/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/mikro-orm/mikro-orm/discussions) or join our [Discord](https://discord.gg/w8bjxFHS7X).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

**Comment by B4nan** [comment]
That relation is not populated.

If you change the line which loads the entities to populate it, it behaves exactly as you want:

```ts
const [user1, user2] = await repoUser.findAll({ populate: ['role.group'] });
```

Btw you are not supposed to be using `EntitySerializer`, we have a global `serialize()` function, or alternatively, you can do `wrap(entity).serialize()`.

**Reply by leonardbinet** [reply]
indeed my bad, I didn't notice the relations weren't populated on this test 🤦 .. I'll let you know if I manage to make a repro of the issue I encounter (where relation is indeed populated).

**Reply by B4nan** [reply]
And you use explicit serialization too?

**Reply by leonardbinet** [reply]
indeed, I've switched to the global `serialize` function though (instead of `EntitySerializer `)

**Reply by leonardbinet** [reply]
in this case, I write `serialize(entity, {populate: ["*"]})`

which might be the bug, note: the reason we use explicit serialization is related to nestjs non-serialization of relations: cf https://github.com/mikro-orm/nestjs/discussions/65, we've implemented this pattern: https://github.com/mikro-orm/nestjs/pull/38/files

**Reply by B4nan** [reply]
> which might be the bug

I doubt it, the only difference there is you can pass in an array of entities to `serialize()`, otherwise its basically an alias:

https://github.com/mikro-orm/mikro-orm/blob/master/packages/core/src/serialization/EntitySerializer.ts#L339

>  is related to nestjs non-serialization of relations

Not sure what you mean by that, that discussion is about missing docs, not about "non-serialization of relations". You are not supposed to be using class-transformer on the entities, if that is what you mean, and nothing changed about that, class-transformer won't understand strongly connected entity graph with cycles at all, neither the `Collection` and `Reference` wrappers.

**Reply by leonardbinet** [reply]
Sorry, my comments are messy.

I'll split the different topics:
- the specific issue I've encountered which I have some difficulties replicating in a simple repo
- nestjs serialization with mikro-orm, for which I'll first dive into other discussions to understand how we could better handle our usecase

### Specific serialization issue

I initially thought that my serialization issue was related to embeddables, but after failing to replicate I'm actually not sure at all.

What I can observe using breakpoints while debugging:
- my entity is properly loaded with all expected relations
- after serialization, some relations are not serialized

<img width="903" alt="Screenshot 2024-10-22 at 10 51 04" src="https://github.com/user-attachments/assets/ecf23c2e-4ff3-405e-9964-9af4277ebfa8">
<img width="1180" alt="Screenshot 2024-10-22 at 10 51 11" src="https://github.com/user-attachments/assets/8cc7564b-aa50-4f06-8ee4-8e97d7e633c6">


At this point, failing to replicate on a repro I can only has suspicions about why it happens here, and nowhere else in our codebase. Here are some elements of context:
- this is the endpoint where we load the largest number of relations: 
```
populate: [
          'contributingDataPoint.tagValues.dataPointTag',
          'contributingDataPoint.unitEntity',
          'contributingDataPoint.source.manualCreator',
          'contributingDataPoint.source.import.creator',
          'contributingDataPoint.source.surveyRecipient',
          'contributingDataPoint.source.extrapolationConfiguration',
          'contributingDataPoint.source.campaignTask.campaign',
          'contributingDataPoint.source.campaignTask.questionnaire',
          'contributingDataPoint.source.campaignTask.assignedContributor',
          'contributingDataPoint.source.campaignFlow.questionnaireToEmissionProcessingRule', // <- I observe that this relation is properly populated, but not serialized
          'contributingDataPoint.source.fileImportFlow.importFileSheet.importFile',
          'contributingEntity',
          'contributingSource',
          'emissionFactor.referenceUnitEntity',
          'dataOrigin',
          'category',
          'supplier',
          'import',
          'campaign',
          'questionnaire',
          'metric.unitGroup'
        ]
```
- the `contributingDataPoint.source` is an embeddable, and downstream populated relations are entities
- I've encountered this issue while migrating from v5 to v6, this endpoint worked in v5


What I now suspect (but that's really a suspicion and it might not be worth much) is that this issue might be related to the redundancy of paths in those relations 
ie
```
          'contributingDataPoint.source.extrapolationConfiguration',
          'contributingDataPoint.source.campaignTask.campaign',
          'contributingDataPoint.source.campaignTask.questionnaire',
          'contributingDataPoint.source.campaignTask.assignedContributor',
          'contributingDataPoint.source.campaignFlow.questionnaireToEmissionProcessingRule'
```
which might not go well with the `packages/core/src/serialization/EntityTransformer.ts:76` cycle detection.

If I can't manage to provide a repro I'll probably try to find a workaround.

**Reply by B4nan** [reply]
Now that's a crazy list of populate hints. If you manage to reproduce this, be sure to heavily simplify, since this wouldn't be debuggable :]

> which might not go well with the packages/core/src/serialization/EntityTransformer.ts:76 cycle detection.

`EntityTransformer` is implicit serialization, not explicit.

**Reply by leonardbinet** [reply]
FYI I used a workaround to populate less relations (using a resolved field in gql to load `contributingDataPoint.source` downstream relations). Loading all those deeply nested relations seems to be the culprit.
